### PR TITLE
[Merged by Bors] - feat: port Algebra.GroupTheory.EckmannHilton

### PR DESCRIPTION
**Pull request by riccardobrasca**
mathlib3 sha: 71dbd401af609bb5065d060b8f221d2a31e2b8d7

### CODE REVIEW

File: Mathlib/Algebra/GroupTheory/EckmannHilton.lean [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+/-- If a type carries a unital magma structure that distributes over a unital binary
+operation, then the magma structure is a commutative monoid. -/
+@[reducible,
+  to_additive
+      "If a type carries a unital additive magma structure that distributes over\na unital binary
+      operation, then the additive magma structure is a commutative additive monoid."]
+def CommMonoid [h : MulOneClass X]
+    (distrib : ∀ a b c d, ((a * b) <m₁> c * d) = (a <m₁> c) * b <m₁> d) : CommMonoid X :=
+  { h with mul := (· * ·), one := 1, mul_comm := (mul_comm h₁ MulOneClass.IsUnital distrib).comm, mul_assoc := (mul_assoc h₁ MulOneClass.IsUnital distrib).assoc }
```
Comment (riccardobrasca) [inline Mathlib/Algebra/GroupTheory/EckmannHilton.lean:95]: How can I fix this? If I put `mul_assoc := ...` in a new line, I get an error "Expected }"

File: Mathlib/Algebra/GroupTheory/EckmannHilton.lean [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+/-- If a type carries a group structure that distributes over a unital binary operation,
+then the group is commutative. -/
+@[reducible,
+  to_additive
+      "If a type carries an additive group structure that\ndistributes over a unital binary
+      operation, then the additive group is commutative."]
+def commGroup [G : Group X]
+    (distrib : ∀ a b c d, ((a * b) <m₁> c * d) = (a <m₁> c) * b <m₁> d) : CommGroup X :=
+  { EckmannHilton.CommMonoid h₁ distrib, G with }
```
Comment (riccardobrasca) [inline Mathlib/Algebra/GroupTheory/EckmannHilton.lean:106]: Is `{  .., G with }` the correct way of writing this? The final `G with` seems wrong.

File: Mathlib/Algebra/GroupTheory/EckmannHilton.lean (hunk 1 above)
Comment (semorrison) [reply Mathlib/Algebra/GroupTheory/EckmannHilton.lean:95]: You have to put `mul := ` on a new line too. It appears to be the rule is either all one line, or each field gets a new line.

File: Mathlib/Algebra/GroupTheory/EckmannHilton.lean (hunk 2 above)
Comment (semorrison) [reply Mathlib/Algebra/GroupTheory/EckmannHilton.lean:106]:
```
  { EckmannHilton.CommMonoid h₁ distrib, G with .. }
```
would be more readable.

**Comment by semorrison** [comment]
Please fix long lines, and add the import line to `Mathlib.lean`. Then probably good to go.

File: Mathlib/GroupTheory/EckmannHilton.lean [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+In fact, the two operations are the same, and give a commutative monoid structure,
+see `eckmann_hilton.CommMonoid`. -/
+theorem one : e₁ = e₂ := by
+  simpa only [h₁.left_id, h₁.right_id, h₂.left_id, h₂.right_id] using distrib e₂ e₁ e₁ e₂
+#align eckmann_hilton.one EckmannHilton.one
+
+/-- If a type carries two unital binary operations that distribute over each other,
+then these operations are equal.
+
+In fact, they give a commutative monoid structure, see `eckmann_hilton.comm_monoid`. -/
```
Comment (pechersky) [inline Mathlib/GroupTheory/EckmannHilton.lean:62]: Here and below, `CommMonoid` instead.

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/GroupTheory/EckmannHilton.lean [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  ⟨fun a b c => by simpa [mul h₁ h₂ distrib, h₂.left_id, h₂.right_id] using distrib a b e₂ c⟩
+#align eckmann_hilton.mul_assoc EckmannHilton.mul_assoc
+
+/-- If a type carries a unital magma structure that distributes over a unital binary
+operation, then the magma structure is a commutative monoid. -/
+@[reducible,
+  to_additive
+      "If a type carries a unital additive magma structure that distributes over\na unital binary
+      operation, then the additive magma structure is a commutative additive monoid."]
+def CommMonoid [h : MulOneClass X]
```
[the remark this answers is not in the document]
Comment (riccardobrasca) [reply Mathlib/GroupTheory/EckmannHilton.lean:93]: Hmm, I think you're right, let me fix it everywhere.

**Comment by semorrison** [comment]
Could you click resolve on the queries, and mark back as `awaiting-review`, when ready? (I think, from the commit messages, that this may be the case already, but don't really want to verify.)

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**